### ISSUE TITLE
Question: transport in separate process without pipe

### ISSUE THREAD

**Issue by sventschui** [pinojs/pino#794]
We are using distroless node.js images (gcr.io/distroless) to run our node.js applications in production. We want to use an Elasticsearch cluster to gather our logs. Pino transports should be offloaded to a separate process for performance reasons. While the docs suggest to use a bash pipe (`|`) to direct stdout of the node.js app into the transport this is not possible in a distroless image as it doesn't include any shell.

The `pino-elasticsearch` module allows to be used as a module but I suspect this is not ideal from a performance point of view. While looking for a multi-threaded solution the following idea came to my mind and I'd like to collect some feedback: The main entrypoint of the docker image is a node.js app that launches the real app as a child_process and uses `pino-elasticsearch` to pump its stdout into elasticsearch. Are there any other caveats other than having to deal properly with a child_process?

**Comment by jsumners** [comment]
At this point, I think it's a better idea to work with your container host ship logs from your application's `stdout`. For example, in AWS ECS Fargate it is possible to start a sibling container that receives the `stdout` from the primary process and ships those logs wherever you like; see https://aws.amazon.com/about-aws/whats-new/2019/11/aws-launches-firelens-log-router-for-amazon-ecs-and-aws-fargate/ .

**Comment by sventschui** [comment]
Sorry for giving to few details in my initial question. For sure it would be the best when the cluster the container is running in would support something like this. Unfortunately I'm currently working at a big insurance company and we're stuck on an on-prem openshift cluster whose EFK stack is unusable (no log parsing, poor perf, ...). Also k8s/openshift does not support reading stdout of another container unless writing stdout to a file what I guess is a bad idea as it would involve log rotation, etc.. Also it would prevent the use of the built in openshift log viewing in the browser UI.

So to sum up I'm pretty stuck with the only option of handling it as part of the application. I see two options:

1. Use `pino-elasticsearch` as a module in the "main" process
2. The proposed solution of having the main process for log forwarding only and running the app as a child process.
3. Move to something other than distroless that includes a shell and allows to pipe/tee the stdout.

I wonder how bad the `pino-elasticsearch` is for perf when running on the main thread. I guess we'd need to test that but I suspect it has quite some impact.

**Comment by jsumners** [comment]
> Also k8s/openshift does not support reading stdout of another container unless writing stdout to a file what I guess is a bad idea as it would involve log rotation, etc..

This would be my preferred option. It's far easier to use standard system tools for rotating and shipping logs.

But, if you want to run your main process in a subprocess just to capture the logs, there's nothing wrong with that. I'd just give the container two cores to work with.

**Comment by sventschui** [comment]
Running the main process as a child process only benefits perf when there is >1 core available. On the other hand in the case of a crash of the main process the log forwarding process will be able to send logs to ES before terminating, that's why I'll probably go that route. Thanks for the discussion! Really helped a lot!
